matrix/app_server/vision/perception_encoder.py

```diff
@@ -0,0 +1,269 @@
+import json
+import logging
+import os
+from argparse import ArgumentParser
+from typing import Any, Dict, List, Optional, Union
+
+import core.vision_encoder.pe as pe
+import core.vision_encoder.transforms as transforms
+import torch
+import torchvision.transforms.v2 as T
+from fastapi import FastAPI
+from PIL import Image
+from ray import get_runtime_context, serve
+from starlette.requests import Request
+from starlette.responses import JSONResponse
+from torch.utils.data import DataLoader, Dataset
+
+from matrix.app_server.vision.utils import (
+    SamplingMode,
+    TorchCodecVideoDataset,
+    get_image_transform,
+)
+
+logger = logging.getLogger("ray.serve")
+logging.getLogger("httpx").setLevel(logging.WARNING)
+logging.getLogger("openai._base_client").setLevel(logging.WARNING)
+
+TIMEOUT = 10
+
+app = FastAPI()
+
+
+def _custom_collate_fn(batch: List[Dict[str, Any]]) -> Dict[str, Any]:
+    """
+    Custom collate function to handle batching of video tensors and metadata.
+    """
+    # Stack all the video tensors into a single batch tensor
+    video_tensors = torch.stack([item["frames"] for item in batch], dim=0)
+
+    # Keep the metadata as a simple list of dictionaries
+    metadata_list = [item["meta"] for item in batch]
+
+    return {
+        "frames": video_tensors,
+        "meta": metadata_list,  # This is now a list of dicts, not a dict of lists/tensors
+    }
+
+
+@serve.deployment(
+    autoscaling_config={
+        "min_replicas": 1,
+        "target_ongoing_requests": 64,
+    },
+    ray_actor_options={"num_cpus": 10, "num_gpus": 1},
+    max_ongoing_requests=100,
+)
+@serve.ingress(app)
+class PerceptionEncoderDeployment:
+    def __init__(
+        self,
+        model_name: str,
+        torch_batch_size: Optional[int] = None,
+    ):
+        self.model_name = model_name
+        self.device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        self.model = (
+            pe.CLIP.from_config(self.model_name, pretrained=True).to(self.device).eval()
+        )
```

Comment:
will torch compile be faster?
self.model = torch.compile(self.model, mode="reduce-overhead")  # or "max-autotune"

Reply:
Done.